[RDY] Eclipse plugin boilerplate

Hjdskes opened the pull request:
Quite a bit had to be restructured for the Eclipse plugin to fit. It's best if this is merged early on, such that we avoid conflicts.

Hjdskes committed:
shell.eclipse: provide boilerplate configuration

I had to change the version numbers in all modules, as the version between the
Eclipse plugin and the Maven module needs to be identical. Eclipse plugins
*have* to use major.minor.micro.qualifier, so the Maven modules versions had to
change.

Hjdskes committed:
Restructure modules

Balletie committed:
Split ReplTest into ConsoleReplTest and a new ReplTest in core

Because Repl was moved to core, ReplTest should test the Repl
generically with mock editor and displays since there is no Terminal UI.

We then let ConsoleReplTest be an integration test with a Terminal UI
and simulated input.

ConsoleReplTest extends from ReplTest and only changes the test setup.
ReplTest was brought into scope by building test-jars in the core
pom.xml, and adding a dependency in the console client pom.xml on the
test jar.

Hjdskes committed:
Move hooks to separate package

Hjdskes commented on shell.core/src/main/java/org/metaborg/spoofax/shell/client/Repl.java:
```diff
@@ -47,6 +37,7 @@
         this.display = display;
         this.invoker = invoker;
 
+        // TODO: this does not really belong here.
```
@Balletie @gfokkema Please discuss. I don't think it belongs here as this class is more of a library kind and should therefore not control the ui. It's better to have the `IEditor` and `IDisplay` implementations set defaults.

Balletie replied:
yea I agree. In fact we are still not sure whether `IEditor` should have a `setPrompt` and `setContinuationPrompt` method.

gfokkema replied:
The reason i moved these here is that in my opinion the prompt should look the same across different implementations for viewers. It is then up to the viewer to decide in what way to convey this look to the user.

I'm afraid that setting different defaults in each implementation could confuse the user as to what shell they're using.

That being said, I also think there's nothing wrong with exposing hooks to shell.console / shell.eclipse that allow the user to change the look of the prompt :p

edit: I'm also the one that moved these methods from the class to the interface in the first place, which at that time was a minor point of discussion, where Skip said: I don't really care right now, we'll come to that later :p

Hjdskes replied:
@Balletie Where else would you move these? A suggestion I made today concerning multiline editing in Eclipse was to just have `enter` go to the next line and using `alt + enter` to send the input to the REPL. This would alleviate the need for a prompt in this editor completely, such that it can be an implementation detail of `TerminalUserInterface`, which does require a prompt. Thoughts?

@gfokkema Why should the prompt be the same if the complete user interface is different? I think that every implementation of `IEditor` and `IDisplay` knows best which environment/user interface it is in and is therefore best knowledgeable to decide its, for example, prompt.

gfokkema replied:
> yea I agree. In fact we are still not sure whether IEditor should have a setPrompt and setContinuationPrompt method.

<break>

> That being said, I also think there's nothing wrong with exposing hooks to shell.console / shell.eclipse that allow the user to change the look of the prompt :p

To elaborate, this is the reason I moved these methods to the interface in the first place: setting a default in Repl to get a uniform look, allowing users to recognize Spoofax Shell by its prompt, and then exposing a function to change these defaults.

gfokkema replied:
After discussing viewpoints and repercussions of moving these methods, I'm in favor of moving this from this class and to change the methods `setPrompt` and `setContinuationPrompt` in the interface of IEditor to default methods `getPrompt` and `getContinuationPrompt`.

Balletie replied:
@gfokkema and then an IEditor would use these methods internally somewhere in their methods, right?

Sounds good to me. Then each subclass can override if needed (but rather not, to keep a uniform look)

Unia commented:
![uml-design](https://cloud.githubusercontent.com/assets/1260727/15227516/e4ceff06-1887-11e6-99bf-ce502f6699be.png)

That's an up-to-date UML of the current changes. Note that the Eclipse module and packages are absent (they are currently being implemented and will be in a separate PR). If it's oke with you, I'd also like to change the groupId and artifactId of our modules to more closely resemble Spoofax.

Hjdskes committed:
Update artifactId and groupId

We now resemble Spoofax upstream more closely. This change also helps in
properly naming the Eclipse plugin and the core module's OSGi bundle.

Unia commented:
I have updated our modules' artifactId and groupId and am now attempting to resolve the build error introduced in 04de04a.

Hjdskes committed:
StyledText: move it from commands to core package

As the updated UML shows (https://cloud.githubusercontent.com/assets/1260727/15242017/ca119324-18f3-11e6-8a3c-1d4fdae7997e.png),
too many classes/packages depend on it. It is best placed into the core package
for this reason.

Hjdskes committed:
Console: fix build

Hjdskes committed:
Fix Unused Import-Package instruction

Unia commented:
I moved the discussion on the `Repl` class setting UI defaults to a new issue (#13).

That means this PR is ready for review now. @Balletie @gfokkema

Unia commented:
I pushed the updated UML to our doc repo.
